### PR TITLE
Overwrite authority if creds servername is specified

### PR DESCRIPTION
**Pull request by menghanl**

### CODE REVIEW

**Comment by tamird** [comment]
test?

**Comment by menghanl** [comment]
Done

File: clientconn_test.go [hunk 1]
```diff
@@ -69,6 +69,22 @@ func TestTLSDialTimeout(t *testing.T) {
...
+		t.Fatalf("Failed to create credentials %v", err)
+	}
+	conn, err := Dial("Non-Existent.Server:80", WithTransportCredentials(creds), WithTimeout(time.Millisecond))
+	if err != nil {
+		t.Fatalf("grpc.Dial(_, _) = _, %v, want _, <nil>", err)
+	}
+	if conn.authority != overwriteServerName {
+		t.Fatalf("conn.authority = %v, want %v", conn.authority, overwriteServerName)
+	}
+	conn.Close()
```
Comment (tamird) [inline clientconn_test.go]: move this after the error check and `defer` it to prevent leaking goroutines when this test fails

Comment (menghanl) [inline clientconn_test.go]: done

**Comment by tamird** [comment]
lgtm

File: credentials/credentials.go [hunk 2]
```diff
@@ -72,14 +72,16 @@ type PerRPCCredentials interface {
...
-// security protocol, security protocol version in use, etc.
+// security protocol, security protocol version in use, server name, etc.
 type ProtocolInfo struct {
 	// ProtocolVersion is the gRPC wire protocol version.
 	ProtocolVersion string
 	// SecurityProtocol is the security protocol in use.
 	SecurityProtocol string
 	// SecurityVersion is the security protocol version.
 	SecurityVersion string
+	// ServerName is ServerName specified by user.
```
Comment (iamqizhao) [inline credentials/credentials.go]: ServerName is the user-specified server name.

File: clientconn_test.go [hunk 3]
```diff
@@ -69,6 +69,22 @@ func TestTLSDialTimeout(t *testing.T) {
...
+	if err != nil {
+		t.Fatalf("Failed to create credentials %v", err)
+	}
+	conn, err := Dial("Non-Existent.Server:80", WithTransportCredentials(creds), WithTimeout(time.Millisecond))
+	if err != nil {
+		t.Fatalf("grpc.Dial(_, _) = _, %v, want _, <nil>", err)
+	}
+	conn.Close()
+	if conn.authority != overwriteServerName {
+		t.Fatalf("conn.authority = %v, want %v", conn.authority, overwriteServerName)
```
Comment (iamqizhao) [inline clientconn_test.go]: "%v.authority ...", conn, ....

File: clientconn_test.go [hunk 4]
```diff
@@ -69,6 +69,22 @@ func TestTLSDialTimeout(t *testing.T) {
...
 
+func TestTLSServerNameOverwrite(t *testing.T) {
+	overwriteServerName := "over.write.server.name"
+	creds, err := credentials.NewClientTLSFromFile(tlsDir+"ca.pem", overwriteServerName)
+	if err != nil {
+		t.Fatalf("Failed to create credentials %v", err)
+	}
+	conn, err := Dial("Non-Existent.Server:80", WithTransportCredentials(creds), WithTimeout(time.Millisecond))
+	if err != nil {
+		t.Fatalf("grpc.Dial(_, _) = _, %v, want _, <nil>", err)
```
Comment (iamqizhao) [inline clientconn_test.go]: grpc.Dial ---> Dial

**Pull request merged**